Additional libtrellis cmake fixes

gsomlo opened the pull request:
I'd like to suggest a few additional follow-up fixes on top of commit 7cd5606, in order to install prjtrellis as a proper, self-contained package (and to make my Fedora packaging work easier in the future -- see issue #56). Three patches are included:

1: Several of the prjtrellis example makefiles reference "tools/bit_to_svf.py".  We should include the "tools" directory into the shared install folder.
2: Distros would frown upon installing binary .so files (e.g., pytrellis.so) under /usr/share/package-name. While we're at it, let's pick the appropriate arch-specific library directory (lib vs. lib64), and install both .so files under a package-specific private subdirectory thereof. This would also affect https://github.com/YosysHQ/nextpnr/pull/210.
3. Finally, we should use /usr/share/packagename as the shared install directory. Right now that's /usr/lib/trellis, but the package name is "prjtrellis", so I'd like to suggest /usr/lib/prjtrellis. This last patch has a strong aesthetic component to it, so consider it an RFC unless everyone agrees it should be applied :)

daveshah1 commented on examples/ecp5_evn/Makefile:
```diff
@@ -6,15 +6,15 @@ all: ${PROJ}.bit
 
 %_out.config: %.json
-	nextpnr-ecp5 --json $< --basecfg ../../misc/basecfgs/empty_lfe5um5g-85f.config --textcfg $@ --um5g-85k --package CABGA381 --lpf ecp5evn.lpf
+	nextpnr-ecp5 --json $< --basecfg /usr/share/prjtrellis/misc/basecfgs/empty_lfe5um5g-85f.config --textcfg $@ --um5g-85k --package CABGA381 --lpf ecp5evn.lpf
```
Please keep the relative paths as they work for everyone regardless of how/where the tools are installed, long term plan is to remove the baseconfigs altogether.

daveshah1 commented on examples/picorv32_versa5g/Makefile:
```diff
@@ -14,16 +14,16 @@ attosoc.json: io_wrapper.v attosoc.v picorv32.v firmware.hex
 
 %.svf: %.bit
-	../../tools/bit_to_svf.py $< $@
+	/usr/share/prjtrellis/tools/bit_to_svf.py $< $@
```
Ditto, although an even nicer option here is to not use the script at all and generate the SVF at the same time as the bitstream with `ecppack --svf`

daveshah1 commented on libtrellis/tools/ecpunpack.cpp:
```diff
@@ -15,7 +15,7 @@ int main(int argc, char *argv[])
     namespace po = boost::program_options;
 
-    std::string database_folder = TRELLIS_PREFIX "/share/trellis/database";
+    std::string database_folder = TRELLIS_PREFIX "/share/prjtrellis/database";
```
I'd prefer to keep this path as trellis, now that many people have it installed already. Otherwise people will end up with two copies of everything which will be confusing.

daveshah1 commented on libtrellis/CMakeLists.txt:
```diff
@@ -108,16 +108,23 @@ endif()
     set_target_properties(${name} PROPERTIES
                           BUILD_WITH_INSTALL_RPATH ON
-                          INSTALL_RPATH "@loader_path/../lib"
+                          INSTALL_RPATH "@loader_path/../${LIBDIR}/prjtrellis"
```
I'd personally prefer to avoid the subfolder and any potential complications it might bring, unless it is a distro requirement.

gsomlo commented:
On Mon, Feb 04, 2019 at 11:34:46PM -0800, David Shah wrote:
> @daveshah1 requested changes on this pull request.
> 
> ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
> 
> In examples/ecp5_evn/Makefile:
> 
> > @@ -6,15 +6,15 @@ all: ${PROJ}.bit
>         yosys -p "synth_ecp5 -noccu2 -nobram -nomux -json $@" $<
> 
>  %_out.config: %.json
> -       nextpnr-ecp5 --json $< --basecfg ../../misc/basecfgs/empty_lfe5um5g-85f.config --textcfg $@ --um5g-85k --package CABGA381 --lpf ecp5evn.lpf
> +       nextpnr-ecp5 --json $< --basecfg /usr/share/prjtrellis/misc/basecfgs/empty_lfe5um5g-85f.config --textcfg $@ --um5g-85k --package CABGA381 --lpf ecp5evn.lpf
> 
> Please keep the relative paths as they work for everyone regardless of how/
> where the tools are installed, long term plan is to remove the baseconfigs
> altogether.

That assumes one has write permissions in the location where the
examples were installed. In a distro package, that usually ends up
being /usr/share/[docs/]<packagename>/... and regular users couldn't
just run "make" in there. Once they copy the example folder to their
home directories, all bets are off w.r.t. relative paths.

If that doesn't sound convincing, I can keep this patch local to the
downstream package (particularly given the planned removal of
baseconfigs) :)
 
> In examples/picorv32_versa5g/Makefile:
> 
> >
>  attosoc.bit: attosoc_out.config
>         ecppack $< $@
> 
>  %.svf: %.bit
> -       ../../tools/bit_to_svf.py $< $@
> +       /usr/share/prjtrellis/tools/bit_to_svf.py $< $@
> 
> Ditto, although an even nicer option here is to not use the script at all and
> generate the SVF at the same time as the bitstream with ecppack --svf

Let me take a stab at rewriting everything with that in mind, new
version of the patch coming soon...
 
> In libtrellis/tools/ecpunpack.cpp:
> 
> > @@ -15,7 +15,7 @@ int main(int argc, char *argv[])
>      using namespace Trellis;
>      namespace po = boost::program_options;
> 
> -    std::string database_folder = TRELLIS_PREFIX "/share/trellis/database";
> +    std::string database_folder = TRELLIS_PREFIX "/share/prjtrellis/database";
> 
> I'd prefer to keep this path as trellis, now that many people have it installed
> already. Otherwise people will end up with two copies of everything which will
> be confusing.

OK, in which case I'll just name the Fedora packages 'trellis[-devel]',
to avoid "squatting" over both 'trellis' and 'prjtrellis' in the
namespace of all possible packages :)
 
> In libtrellis/CMakeLists.txt:
> 
> >  function(setup_rpath name)
>    if(APPLE)
>      set_target_properties(${name} PROPERTIES
>                            BUILD_WITH_INSTALL_RPATH ON
> -                          INSTALL_RPATH "@loader_path/../lib"
> +                          INSTALL_RPATH "@loader_path/../${LIBDIR}/prjtrellis"
> 
> I'd personally prefer to avoid the subfolder and any potential complications
> it might bring, unless it is a distro requirement.

I can leave the "if(APPLE)" branch of setup_rpath alone if that's a
concern. However, you're *already* configuring rpath for the
executables, and yes, Fedora wants it to be package-specific in this
type of situation (arch-specific .so DLLs that are not versioned and
only intended for use by binaries within the package they're part of).
See here for details:

https://lists.fedoraproject.org/pipermail/devel/2012-April/166104.html
https://fedoraproject.org/wiki/Common_Rpmlint_issues#invalid-soname
https://fedoraproject.org/wiki/Common_Rpmlint_issues#no-soname

These are Fedora specific, but probably a good idea elsewhere as well.

I can always keep patches with the downstream package if they're total
dealbreakers for upstream, but figured at least I'd try to push the
ones that obviously feel like they might be "good ideas" before going
that route.

Thanks much,
--Gabriel

> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.*
>

daveshah1 commented:
> That assumes one has write permissions in the location where the examples were installed. In a distro package, that usually ends up being /usr/share/[docs/]<packagename>/... and regular users couldn't just run "make" in there. Once they copy the example folder to their home directories, all bets are off w.r.t. relative paths. If that doesn't sound convincing, I can keep this patch local to the downstream package (particularly given the planned removal of baseconfigs) :)

Makes sense. Maybe a make variable called TRELLIS defaulting to /usr/share/trellis would be a good solution?

> I can leave the "if(APPLE)" branch of setup_rpath alone if that's a concern. However, you're *already* configuring rpath for the executables, and yes, Fedora wants it to be package-specific in this type of situation (arch-specific .so DLLs that are not versioned and only intended for use by binaries within the package they're part of). See here for details: 

Thanks for the links. Subfolder seems a good solution them.

gsomlo commented:
I dropped the patch that tried to rename install paths to "prjtrellis" and adjust all relative paths across the included examples.

I added a patch that generates .svf from .config directly (using `ecppack --svf`).

I'll hold off on messing with relative paths inside example Makefiles for a separate PR (to be submitted after https://github.com/YosysHQ/nextpnr/pull/228 lands).

gsomlo commented:
Hmm, now that bit_to_svf.py is no longer in play, maybe I should also drop the first patch that installs the tools folder?

daveshah1 commented:
Thanks for your work on this!

se-bi commented:
What was the specific reason for this commit ba8fa459758752a9675bdd6792847ec59aed0096 (_Preserve Python script execute permissions during cmake install_) within this PR  
or rather to not apply this on the line before / similar directory with python scripts?

If one builds the projects as a user and applies `umask 022` before `sudo make install`,  
the one directory with the `USE_SOURCE_PERMISSIONS` has still different access permissions than the others,  
unless this project was cloned with `umask 022`, which is inconsistent and inconvenient.

I would suggest revert that commit ba8fa459758752a9675bdd6792847ec59aed0096 !?

gsomlo commented:
On Sun, Oct 18, 2020 at 05:11:52AM -0700, se-bi wrote:
> What was the specific reason for this commit ba8fa45 (Preserve Python script
> execute permissions during cmake install) within this PR
> or rather to not apply this on the line before / similar directory with python
> scripts?

I wish this came up back when it was fresh in my mind :) Looking back,
the main difference that may explain it is that .py scripts in the
timing folder have a python shebang (and are therefore meant to work
as standalone executables), whereas the ones in the util/common
folder do not (and must be invoked via `python[3] foo.py`). Hence the
need to preserve executable flags on the former, but not the latter.

se-bi commented:
> the main difference that may explain it is that .py scripts in the timing folder have a python shebang (and are therefore meant to work as standalone executables), whereas the ones in the util/common folder do not (and must be invoked via `python[3] foo.py`). Hence the need to preserve executable flags on the former, but not the latter.

That correlates then with the execution bit set to allow the standalone execution

Independent from that, the option `USE_SOURCE_PERMISSIONS` keeps the permission `[67]00` file permission when using `sudo make install`,
after the user has cloned with a default `umask 077` and applies `umask 022` before `sudo make install`
(or is that an usual setting and procedure)

Without that option, at least the read permission is set for group and other, in the previous named scenario.

Also, in that scenario the _nextpnr_ build fails without adapting the permission for that specific directory,
those files remain only readable for the user, which is then root.

gsomlo commented:
> Without that option, at least the read permission is set for group and other,
> in the previous named scenario.
> 
> Also, in that scenario the nextpnr build fails without adapting the permission
> for that specific directory,
> those files remain only readable for the user, which is then root.

My (vanilla) umask as a regular Fedora 32 user is 022. I (recursively)
cloned the prjtrellis repo -- without bringing up the umask in any way,
just:

```
git clone .../prjtrellis
cd prjtrellis
git submodule update --init --recursive
cmake libtrellis
make
sudo make install
```

I ended up with the following (showing only the folder under question,
the one with USE_SOURCE_PERMISSIONS applied to it in the cmakefile):

```
ls -alh /usr/local/share/trellis/timing/util/
total 64K
drwxr-xr-x. 2 root root 4.0K Oct 20 16:36 .
drwxr-xr-x. 3 root root 4.0K Oct 20 16:36 ..
-rw-r--r--. 1 root root 1.6K Oct 20 16:27 cell_fuzzers.py
-rwxr-xr-x. 1 root root 5.5K Oct 20 16:27 cell_html.py
-rwxr-xr-x. 1 root root 3.4K Oct 20 16:27 cell_timings.py
-rw-r--r--. 1 root root 2.2K Oct 20 16:27 design_pip_classes.py
-rw-r--r--. 1 root root 6.3K Oct 20 16:27 extract_ncl_routing.py
-rw-r--r--. 1 root root   20 Oct 20 16:27 .gitignore
-rw-r--r--. 1 root root    0 Oct 20 16:27 __init__.py
-rw-r--r--. 1 root root 2.2K Oct 20 16:27 interconnect_html.py
-rw-r--r--. 1 root root 3.8K Oct 20 16:27 parse_sdf.py
-rw-r--r--. 1 root root 5.7K Oct 20 16:27 pip_classes.py
-rw-r--r--. 1 root root  717 Oct 20 16:27 timing_dbs.py
-rw-r--r--. 1 root root 3.6K Oct 20 16:27 timing_solver.py
```

That looks OK to me. Are you getting different results? Are the
results I am getting not OK from your perspective? If so, why? :)

Cheers,
--G

se-bi commented:
Ok, i must have set in my (then not so vanilla) Debian the umask (for my user) to 077,  
therefore I do `umask 022` before `sudo make install`  

Usually that's enough right before the install.  
Here, I can achieve the wanted and your result,
only when I apply the `umask 022` before the clone. (Upper half)

If not, and done later, like for the lower half,
that one directory remains inaccessible for others.
(The execution bit is copied correctly in each case)

Hence, without that `USE_SOURCE_PERMISSIONS` it would work for each user independent of their umask.
It is more about the permissions for the group and others, than about the execution bit.

```
ls -hl /opt/prjtrellis-0*/share/trellis/{util/common,timing/util}
#####
# umask 022 before git clone

/opt/prjtrellis-022-2020-09-26-23d3464/share/trellis/timing/util:
total 52K
-rw-r--r-- 1 root root    0 Oct 20 18:46 __init__.py
-rw-r--r-- 1 root root 1.6K Oct 20 18:46 cell_fuzzers.py
-rwxr-xr-x 1 root root 5.5K Oct 20 18:46 cell_html.py
[..]
-rw-r--r-- 1 root root 5.7K Oct 20 18:46 pip_classes.py
-rw-r--r-- 1 root root  717 Oct 20 18:46 timing_dbs.py
-rw-r--r-- 1 root root 3.6K Oct 20 18:46 timing_solver.py

/opt/prjtrellis-022-2020-09-26-23d3464/share/trellis/util/common:
total 28K
-rw-r--r-- 1 root root    0 Oct 20 18:46 __init__.py
[...]
drwxr-xr-x 2 root root 4.0K Oct 20 19:02 nets
-rw-r--r-- 1 root root  504 Oct 20 18:46 tiles.py

#####
# umask 022 before make install

/opt/prjtrellis-077-2020-09-26-23d3464/share/trellis/timing/util:
total 52K
-rw------- 1 root root    0 Oct 20 18:46 __init__.py
-rw------- 1 root root 1.6K Oct 20 18:46 cell_fuzzers.py
-rwx------ 1 root root 5.5K Oct 20 18:46 cell_html.py
[..]
-rw------- 1 root root 5.7K Oct 20 18:46 pip_classes.py
-rw------- 1 root root  717 Oct 20 18:46 timing_dbs.py
-rw------- 1 root root 3.6K Oct 20 18:46 timing_solver.py

/opt/prjtrellis-077-2020-09-26-23d3464/share/trellis/util/common:
total 28K
-rw-r--r-- 1 root root    0 Oct 20 18:46 __init__.py
[..]
drwxr-xr-x 2 root root 4.0K Oct 20 19:02 nets
-rw-r--r-- 1 root root  504 Oct 20 18:46 tiles.py
```

gsomlo commented:
Have you tried any of this on a fresh install (e.g., on a vm) with
"factory default" umask settings (for *both* your regular user and
root)?

se-bi commented:
> Have you tried any of this on a fresh install (e.g., on a vm) with "factory default" umask settings (for *both* your regular user and root)?

Sure, then it's fine as well.

Sorry, lost track of the original issue, that made that change necessary.

What about this solution (f3baa0e, 
on my fork [se-bi/prjtrellis](https://github.com/se-bi/prjtrellis/tree/fix/install-file-exec-perm), funny though, that github matches this commit as a link in this repo, anyway ... ):
```
- install(DIRECTORY ../timing/util DESTINATION ${CMAKE_INSTALL_DATADIR}/${PROGRAM_PREFIX}trellis/timing USE_SOURCE_PERMISSIONS)
+ install(DIRECTORY ../timing/util DESTINATION ${CMAKE_INSTALL_DATADIR}/${PROGRAM_PREFIX}trellis/timing)
+ install(PROGRAMS ../timing/util/cell_html.py DESTINATION ${CMAKE_INSTALL_DATADIR}/${PROGRAM_PREFIX}trellis/timing/util)
+ install(PROGRAMS ../timing/util/cell_timings.py DESTINATION ${CMAKE_INSTALL_DATADIR}/${PROGRAM_PREFIX}trellis/timing/util)
```

This umask thing reveals the way of ba8fa45, at least when somebody uses by default `077`.
The `install(PROGRAMS ...)` instructs cmake/make to set the exec-bit when copying ([see diff. in `FILES` and `PROGRAMS`](https://cmake.org/cmake/help/latest/command/install.html#installing-files)).
How the group and other permission in general are set depends then on the umask.
(Installs those files kind of twice though ...)

If that works for you as well, (works for me :grin:), i would open a new PR!?

gsomlo commented:
On Fri, Oct 23, 2020 at 11:51:49AM -0700, se-bi wrote:
> 
> If that works for you as well, (works for me 😁), i would open a new PR!?
> 

Can confirm that f3baa0e behaves identically to current upstream in
the "factory default" umask case, so I'd have no personal reason to
object against it. Yes, opening a new PR and letting @daveshah1 make
the final call is probably the right way forward.
